File: backend/app/schemas/pdf.py
from __future__ import annotations

from pydantic import BaseModel, Field

from app.models.job import JobStatus


class PdfGenerationSettings(BaseModel):
    image_quality: int = Field(default=92, ge=70, le=100)
    include_cover: bool = True
    include_topic_dividers: bool = True
    include_context: bool = False


class StartPdfGenerationRequest(BaseModel):
    video_id: str
    settings: PdfGenerationSettings = Field(default_factory=PdfGenerationSettings)


class StartPdfGenerationResponse(BaseModel):
    job_id: str
    video_id: str
    status: JobStatus
    reused_existing: bool
    message: str


class PdfReviewItem(BaseModel):
    position: int
    trusted_index: int
    candidate_index: int
    frame_index: int
    timestamp_seconds: float
    topic_index: int
    topic_title: str
    image_filename: str
    auto_protected: bool
    content_loss_risk: bool


class PdfReviewSummary(BaseModel):
    video_id: str
    pipeline_version: int
    screenshot_count: int
    topic_count: int
    coverage_generated_at: str
    trusted_generated_at: str
    topics_generated_at: str
    updated_at: str


class PdfReviewResponse(BaseModel):
    status: str = "ready"
    summary: PdfReviewSummary
    items: list[PdfReviewItem]


class UpdatePdfReviewRequest(BaseModel):
    ordered_trusted_indexes: list[int]


class PdfArtifact(BaseModel):
    video_id: str
    status: str
    pipeline_version: int
    file_name: str
    file_size_bytes: int
    page_count: int
    screenshot_count: int
    topic_count: int
    settings: PdfGenerationSettings
    coverage_generated_at: str
    review_updated_at: str
    trusted_generated_at: str
    topics_generated_at: str
    generated_at: str


class PdfResultResponse(BaseModel):
    status: str = "ready"
    pdf: PdfArtifact

File: frontend/src/types/pdf.ts
export type PdfJobStatus =
  | "QUEUED"
  | "GENERATING_PDF"
  | "READY"
  | "FAILED"
  | "INTERRUPTED"
  | "CANCELLED";

export interface PdfGenerationSettings {
  image_quality: number;
  include_cover: boolean;
  include_topic_dividers: boolean;
  include_context: boolean;
}

export interface StartPdfGenerationResponse {
  job_id: string;
  video_id: string;
  status: PdfJobStatus;
  reused_existing: boolean;
  message: string;
}

export interface PdfJobResponse {
  job_id: string;
  video_id: string;
  job_type: "PDF_GENERATION";
  status: PdfJobStatus;
  progress: number;
  message: string;
  error: { code: string; message: string } | null;
}

export interface PdfReviewItem {
  position: number;
  trusted_index: number;
  candidate_index: number;
  frame_index: number;
  timestamp_seconds: number;
  topic_index: number;
  topic_title: string;
  image_filename: string;
  auto_protected: boolean;
  content_loss_risk: boolean;
}

export interface PdfReviewSummary {
  video_id: string;
  pipeline_version: number;
  screenshot_count: number;
  topic_count: number;
  coverage_generated_at: string;
  trusted_generated_at: string;
  topics_generated_at: string;
  updated_at: string;
}

export interface PdfReviewResponse {
  status: "ready";
  summary: PdfReviewSummary;
  items: PdfReviewItem[];
}

export interface PdfArtifact {
  video_id: string;
  status: string;
  pipeline_version: number;
  file_name: string;
  file_size_bytes: number;
  page_count: number;
  screenshot_count: number;
  topic_count: number;
  settings: PdfGenerationSettings;
  coverage_generated_at: string;
  review_updated_at: string;
  trusted_generated_at: string;
  topics_generated_at: string;
  generated_at: string;
}

export interface PdfResultResponse {
  status: "ready";
  pdf: PdfArtifact;
}
